internal/terminal/table.go

```diff
@@ -0,0 +1,139 @@
+package terminal
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/fatih/color"
+)
+
+const (
+	logFieldHeaders = "headers"
+	logFieldData    = "data"
+)
+
+type table struct {
+	headers      []string
+	data         []map[string]string
+	columnWidths map[string]int
+}
+
+var (
+	// gutter is to set a standard for the break between columns for the table
+	gutter = strings.Repeat(" ", 2)
+)
+
+var (
+	tableFields = []string{logFieldData, logFieldHeaders}
+)
+
+func newTable(headers []string, data []map[string]interface{}) table {
+	var t table
+
+	if len(headers) == 0 {
+		return t
+	}
+
+	t.headers = headers
+	t.data = make([]map[string]string, len(data))
+	t.columnWidths = make(map[string]int, len(headers))
+
+	for _, header := range headers {
+		t.columnWidths[header] = len(header)
+	}
+
+	for i, row := range data {
+		r := make(map[string]string)
+		for _, header := range t.headers {
+			// convert the interface data into strings for printing
+			value := parseValue(row[header])
+			if width := len(value); width > t.columnWidths[header] {
+				t.columnWidths[header] = width
+			}
+			r[header] = value
+		}
+		t.data[i] = r
+	}
+	return t
+}
+
+func (t table) Message() (string, error) {
+	if err := t.validate(); err != nil {
+		return "", err
+	}
+	return fmt.Sprintf(`
+%s
+%s
+%s`,
+		t.headerString(), t.dividerString(), t.rowString()), nil
+}
+
+func (t table) Payload() ([]string, map[string]interface{}, error) {
+	if err := t.validate(); err != nil {
+		return nil, nil, err
+	}
+	return tableFields, map[string]interface{}{
+		logFieldHeaders: t.headers,
+		logFieldData:    t.data,
+	}, nil
+}
+
+func (t table) headerString() string {
+	cells := make([]string, len(t.headers))
+	for i, header := range t.headers {
+		cells[i] = fmt.Sprintf("%s%s",
+			color.New(color.Bold).SprintFunc()(header),
+			strings.Repeat(" ", t.columnWidths[header] - len(header)),
+		)
+	}
+	return strings.Join(cells, gutter)
+}
+
+func (t table) rowString() string {
```

Comment:
sorry about this, i was being dumb earlier!  I didn't connect the dots this was stringifying all the rows, the name should probably either `rowsString` or `dataString` (I think I'm leaning towards the latter, but your call there).

and then `cells` should probably turn into `rows` instead.... sorry about that :bushes: